File: src/main/java/me/dkim19375/dkim19375core/CoreJavaPlugin.java
package me.dkim19375.dkim19375core;

import org.bukkit.Bukkit;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.PluginCommand;
import org.bukkit.command.TabCompleter;
import org.bukkit.event.Listener;
import org.bukkit.plugin.java.JavaPlugin;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.HashSet;
import java.util.Set;
import java.util.logging.Level;
import java.util.logging.Logger;

public abstract class CoreJavaPlugin extends JavaPlugin {
    private static final Logger logger = Bukkit.getLogger();
    private final Set<ConfigFile> files = new HashSet<>();

    public static void log(Level level, String msg) { logger.log(level, msg); }

    @SuppressWarnings("unused")
    public static void log(String msg) { log(Level.INFO, msg); }

    @Override
    public abstract void onEnable();

    @Override
    public abstract void onDisable();

    @Override
    public void reloadConfig() {
        saveDefaultConfig();
        super.reloadConfig();
        files.forEach(this::reloadConfig);
    }

    public void reloadConfig(ConfigFile config) {
        config.saveDefaultConfig();
        config.reload();
    }

    public void registerCommand(@NotNull String command, @NotNull CommandExecutor executor, @Nullable TabCompleter tabCompleter) {
        final PluginCommand cmd = getCommand(command);
        if (cmd == null) {
            logger.log(Level.SEVERE, "Could not register command: " + command + "!");
            getServer().getPluginManager().disablePlugin(this);
            throw new NullPointerException("[" + getDescription().getName() + "] Could not register command: " + command + "!");
        }
        cmd.setExecutor(executor);
        if (tabCompleter != null) {
            cmd.setTabCompleter(tabCompleter);
        }
    }

    @SuppressWarnings("unused")
    public void registerCommand(@NotNull String command, @NotNull CommandExecutor executor) {
        registerCommand(command, executor, null);
    }

    @SuppressWarnings("unused")
    public void registerListener(Listener listener) {
        getServer().getPluginManager().registerEvents(listener, this);
    }

    @SuppressWarnings("unused")
    public void registerConfig(ConfigFile config) {
        unregisterConfig(config);
        files.add(config);
        reloadConfig(config);
    }

    public void unregisterConfig(ConfigFile config) {
        unregisterConfig(config.getName());
    }

    public void unregisterConfig(String config) {
        new HashSet<>(files).forEach(file -> {
            if (file.getName().equals(config)) {
                files.remove(file);
            }
        });
    }

    public Set<ConfigFile> getRegisteredFiles() {
        return files;
    }
}

File: src/main/kotlin/me/dkim19375/dkim19375core/javaplugin/CoreJavaPlugin.kt
package me.dkim19375.dkim19375core.javaplugin

import me.dkim19375.dkim19375core.annotation.API
import me.dkim19375.dkim19375core.config.ConfigFile
import org.bukkit.command.CommandExecutor
import org.bukkit.command.TabCompleter
import org.bukkit.event.Listener
import org.bukkit.plugin.java.JavaPlugin
import java.util.function.Consumer
import java.util.logging.Level

@API
abstract class CoreJavaPlugin : JavaPlugin() {
    private val files: MutableSet<ConfigFile> = mutableSetOf()

    override fun reloadConfig() {
        saveDefaultConfig()
        super.reloadConfig()
        files.forEach { config: ConfigFile ->
            this.reloadConfig(
                config
            )
        }
    }

    @API
    fun reloadConfig(config: ConfigFile) {
        config.saveDefaultConfig()
        config.reload()
    }

    @API
    fun registerCommand(command: String, executor: CommandExecutor, tabCompleter: TabCompleter?) {
        val cmd = getCommand(command)
            ?: throw IllegalStateException("[" + description.name + "] Could not register command: " + command + "!")
        cmd.setExecutor(executor)
        tabCompleter ?: return
        cmd.tabCompleter = tabCompleter
    }

    @API
    fun registerCommand(command: String, executor: CommandExecutor) = registerCommand(command, executor, null)

    @API
    fun registerListener(vararg listener: Listener) = listener.forEach {
        server.pluginManager.registerEvents(it, this)
    }

    @API
    fun registerConfig(config: ConfigFile) {
        unregisterConfig(config)
        files.add(config)
        reloadConfig(config)
    }

    @API
    fun unregisterConfig(config: ConfigFile) = unregisterConfig(config.fileName)

    @API
    fun unregisterConfig(configFileName: String) {
        HashSet(files).forEach(Consumer { file: ConfigFile ->
            if (file.fileName == configFileName) {
                files.remove(file)
            }
        })
    }

    @API
    fun getRegisteredFiles(): Set<ConfigFile> {
        return files
    }
}